### PR TITLE
For /pun/sys/shell (host not defined), ssh to default_sshhost

### PR DESCRIPTION
**Pull request by matthu017**

### CODE REVIEW

**Comment by msquee** [comment]
This fixes #569 

If `NODE_ENV` is not `production` then you need to add the following to `.env.local`:
```
DEFAULT_SSHHOST=owens.osc.edu
```

File: apps/shell/app.js [hunk 1]
```diff
@@ -78,8 +78,13 @@ default_sshhost = process.env.DEFAULT_SSHHOST || default_sshhost;
...
   let match = url.match(host_path_rx),
-  hostname = match[1] === "default" ? default_sshhost : match[1],
-  directory = match[2] ? decodeURIComponent(match[2]) : null;
+  hostname = default_sshhost, 
+  directory = null;
+
+  if (match) {
+    hostname = match[1] === "default" ? default_sshhost : match[1];
+    directory = match[2] ? decodeURIComponent(match[2]) : null;
+  }
```
Comment (ericfranz) [inline apps/shell/app.js:87]: So if the URL doesn't match for any reason, we use default host and directory null. I don't know why that makes me nervous. Seems to broad. I don't have a better suggestion at this point. @johrstrom any suggestion?

Comment (msquee) [reply apps/shell/app.js:87]: The only way this fails is if there's no default host set in `/etc/ood/config/clusters.d/*.y*ml` We could add additional logic that goes if no default host is set in yaml configs or through an env var then default to use first host in y*ml configs

**Comment by ericfranz** [comment]
Maybe instead we change

```
// Create all your routes
var router = express.Router();
router.get('/', function (req, res) {
  res.redirect(req.baseUrl + '/ssh');
});
```

to redirect to `req.baseUrl + '/ssh/default'` instead. And catch just '/ssh' without anything trailing it to redirect to '`/ssh/default`

File: apps/shell/app.js (hunk 1 above)
Comment (ericfranz) [reply apps/shell/app.js:87]: I like the first host idea. So we would do

   default_sshhost = process.env.DEFAULT_SSHHOST || default_sshhost || first_available_host;

File: apps/shell/app.js [hunk 2]
```diff
@@ -78,8 +78,13 @@ default_sshhost = process.env.DEFAULT_SSHHOST || default_sshhost;
 if (default_sshhost) host_whitelist.add(default_sshhost);
 function host_and_dir_from_url(url){
   let match = url.match(host_path_rx),
-  hostname = match[1] === "default" ? default_sshhost : match[1],
-  directory = match[2] ? decodeURIComponent(match[2]) : null;
+  hostname = default_sshhost,
```
Comment (ericfranz) [inline apps/shell/app.js:81]: So my question below about being broad is really the question of whether in the case that the host regex does not match, do we default the hostname to default_sshhost instead of null? I think if we do the suggestions in the other comments, we can default this to null instead. That way, an incorrectly formed URL errors out gracefully, instead of logging you into the default host.

File: apps/shell/app.js (hunk 1 above)
Comment (matthu017) [reply apps/shell/app.js:87]: The host allowlist is a Set so there is no 'first' host. We can do the first host found when going in the cluster configs although problems may arise when the host is unavailable for general access even if there is a present cluster config (ruby.osc.edu in our case). This is a very specific edge case.
However this shouldn't be a problem if we make the `default: ` field necessary for cluster configs.

File: apps/shell/app.js (hunk 2 above)
Comment (matthu017) [reply apps/shell/app.js]: Submitting `https://ondemand-test.osc.edu/pun/dev/shell/this/url/has/no/host` returns `Cannot GET /pun/dev/shell/this/url/has/no/host`. 
There is no other error reporting that might guide the user to the solution though. Another edge case as the user would have to try to break the url.

File: apps/shell/app.js [hunk 3]
```diff
@@ -72,23 +72,32 @@ glob.sync(path.join((process.env.OOD_CLUSTERS || '/etc/ood/config/clusters.d'),
...
   }
-
   return [hostname, directory];
 }
 
+let foo, bar = null;
+console.log(foo);
+console.log(bar);
+
+
```
Comment (msquee) [inline apps/shell/app.js:100]:
```suggestion
```

File: apps/shell/app.js [hunk 4]
```diff
@@ -72,15 +72,24 @@ glob.sync(path.join((process.env.OOD_CLUSTERS || '/etc/ood/config/clusters.d'),
     let isDefault = config.v2.login.default;
     host_whitelist.add(host);
     if (isDefault) default_sshhost = host;
+    if (isFirst) {
+      first_available_host = host;
+      isFirst = false;
```
Comment (ericfranz) [inline apps/shell/app.js:77]: You don't need an extra boolean, just do:

    if(! first_available_host) first_available_host = host;

File: apps/shell/app.js (hunk 1 above)
Comment (ericfranz) [reply apps/shell/app.js:92]: I like the implementation you added (except for one minor improvement suggested above). The first available host is useful because that will be the case every new site experiences when installing OnDemand and adding their first cluster config. A default option in the login section of the cluster config is a new concept, so no sites will have this set when installing OnDemand, so it shouldn't be required. As for "the host is unavailable for general access" I think we will address that with 2.0 - getting rid of those atrocious Ruby enforced ACLs and asking people to use file permissions on the cluster configs like we do on the apps.
